packages/legend-query-builder/src/stores/__tests__/QueryBuilderChangeHistory.test.ts

```diff
@@ -0,0 +1,95 @@
+import { integrationTest } from '@finos/legend-shared/test';
+import type { Entity } from '@finos/legend-storage';
+import { test } from '@jest/globals';
+import { waitFor, getByText, fireEvent } from '@testing-library/react';
+import { act } from 'react-dom/test-utils';
+import { TEST__setUpQueryBuilder } from '../../components/__test-utils__/QueryBuilderComponentTestUtils.js';
+import { QUERY_BUILDER_TEST_ID } from '../../__lib__/QueryBuilderTesting.js';
+import { QueryBuilderTDSState } from '../fetch-structure/tds/QueryBuilderTDSState.js';
+import { TEST_DATA__ModelCoverageAnalysisResult_ChangeDetection } from './TEST_DATA__ModelCoverageAnalysisResult.js';
+import TEST_DATA__ChangeDetectionModel from './TEST_DATA__QueryBuilder_Model_ChangeDetection.json' assert { type: 'json' };
+import { TEST_DATA__TestChangeDetectionWithSimpleProject } from './TEST_DATA__QueryBuilder_TestChangeDetection.js';
+
+test(integrationTest('Test change detection'), () => {
+  async () => {
+    const { renderResult, queryBuilderState } = await TEST__setUpQueryBuilder(
+      TEST_DATA__ChangeDetectionModel as Entity[],
+      stub_RawLambda(),
+      'my::map',
+      'my::runtime',
+      TEST_DATA__ModelCoverageAnalysisResult_ChangeDetection,
+    );
+
+    await act(async () => {
+      queryBuilderState.changeClass(
+        queryBuilderState.graphManagerState.graph.getClass('my::Firm'),
+      );
+    });
+    const setupPanel = await waitFor(() =>
+      renderResult.getByTestId(QUERY_BUILDER_TEST_ID.QUERY_BUILDER_SETUP),
+    );
+    await waitFor(() =>
+      getByText(setupPanel, extractElementNameFromPath('my::Firm')),
+    );
+    await waitFor(() =>
+      getByText(setupPanel, extractElementNameFromPath('my::map')),
+    );
+    await waitFor(() =>
+      getByText(setupPanel, extractElementNameFromPath('my::runtime')),
+    );
+    await act(async () => {
+      queryBuilderState.initializeWithQuery(
+        create_RawLambda(
+          TEST_DATA__TestChangeDetectionWithSimpleProject.parameters,
+          TEST_DATA__TestChangeDetectionWithSimpleProject.body,
+        ),
+      );
+    });
+    const projectionCols = await waitFor(() =>
+      renderResult.getByTestId(QUERY_BUILDER_TEST_ID.QUERY_BUILDER_TDS),
+    );
+    await waitFor(() => getByText(projectionCols, 'Legal Name'));
+    guaranteeNonNullable(
+      guaranteeType(
+        queryBuilderState.fetchStructureState.implementation,
+        QueryBuilderTDSState,
+      ).projectionColumns[0],
+    ).setColumnName('Name');
+    await waitFor(() => getByText(projectionCols, 'Name'));
+    guaranteeNonNullable(
+      guaranteeType(
+        queryBuilderState.fetchStructureState.implementation,
+        QueryBuilderTDSState,
+      ).projectionColumns[0],
+    ).setColumnName('Legal Name');
+    await waitFor(() => getByText(projectionCols, 'Legal Name'));
+
+    // Test Redo/Undo action in Query Builder
+    fireEvent.click(renderResult.getByText('Undo'));
```

Comment:
This test does not really test undo/redo. 
After first undo, you should ensure you no longer see Legal Name in the projection panel. 

Additionally can you add more complicated operations. 
Maybe DND in filter. 

If the expectation is that undo/redo do not work in modals, can you test this here as well ?

Reply:
`await waitFor(() => getByText(projectionCols, 'Name'));` is checking the projection panel contains 'Name' as getByText looks for an exact match of the provided string within the text content of the elements.
 However, I could add a check to expect Legal Name to be null

Reply:
yes but the way the current test is written, undo/redo could both be broken and the test would still pass as you are loading it with both 'Name' and 'LegalName` and then doing checks on both those columns without checking if the one who shouldnt show is actually gone.